Add 'enum_select' diagnostic selection to clang.

erichkeane opened the pull request:
This causes us to generate an enum to go along with the select diagnostic, which allows for clearer diagnostic error emit lines.

The syntax for this is:

%enum_select<EnumerationName>{%OptionalEnumeratorName{Text}|{Text2}}0

Where the curley brackets around the select-text are only required if an Enumerator name is provided.

The TableGen here emits this as a normal 'select' to the frontend, which permits us to reuse all of the existing 'select' infrastructure. Documentation is the same as well.

erichkeane committed:
Add 'enum_select' diagnostic selection to clang.

This causes us to generate an enum to go along with the select
diagnostic, which allows for clearer diagnostic error emit lines.

The syntax for this is:

%enum_select<EnumerationName>{%OptionalEnumeratorName{Text}|{Text2}}0

Where the curley brackets around the select-text are only required if an
Enumerator name is provided.

The TableGen here emits this as a normal 'select' to the frontend, which
permits us to reuse all of the existing 'select' infrastructure.
Documentation is the same as well.

AaronBallman commented on clang/test/TableGen/select-enum-errors.td:
```diff
@@ -0,0 +1,16 @@
+def DupeValue : Error<"%enum_select<DupeValue>{%Name{V1}|%Name{V2}}0">;
+// CHECK: error: Duplicate enumerator name 'Name'
+
+def EnumValNotExpected : Error<"%enum_select{V1|%Val2{V2}}0">;
```
How do you feel about things like: `%enum_select<?>{%^no^{V1}}0` where there are invalid identifier characters being used? No idea how bad the diagnostic behavior would be falling out of just assuming nobody does that.

erichkeane replied:
I thought about that.  We end up not diagnosing a lot of this sort of thing, so I didn't know how much value there was to it, particularly when continuing to build will cause the error later when the enum fails to parse/compile.  If we have a good function to test for it, I could potentially error on a non-identifier character?

AaronBallman replied:
Eh, let's not bother; I don't imagine this is a problem we'll hit all that often and it should still produce diagnostics somewhere, so if we do, we'll know.

AaronBallman commented on clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:
```diff
@@ -886,6 +909,11 @@ struct DiagTextDocPrinter : DiagTextVisitor<DiagTextDocPrinter> {
   }
 
+  void VisitEnumSelect(EnumSelectPiece *P) {
+    // For now, document this as if it were a 'select'.
```
Should we file an issue to address this?

erichkeane replied:
I don't think so?  Actually, the 'For now,' could probably go away.  The documentation page seems more directed at someone 'reading' the documentation, so the "list of possible things this could say" is likely sufficient.  WDYT?

AaronBallman replied:
I think that makes sense -- let's drop the "for now" in the comment and just explain that in terms of documentation, there's no difference between select and enum_select.

AaronBallman commented on clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:
```diff
@@ -1042,6 +1119,9 @@ Piece *DiagnosticTextBuilder::DiagText::parseDiagText(StringRef &Text,
     };
 
+    if (ModType != MT_EnumSelect && Text[0] == '<')
+      Builder.PrintFatalError("modifier '<' syntax not valid with " + Modifier);
```
Test coverage for this bit?

AaronBallman commented on clang/test/TableGen/select-enum-errors.td:
```diff
@@ -0,0 +1,16 @@
+// RUN: clang-tblgen --gen-clang-diags-enums -I%S %s 2>&1 | FileCheck %s
+include "DiagnosticBase.inc"
+
+def DupeNames1 : Error<"%enum_select<DupeName>{}0">;
```
Other test cases:
```
// Missing closing > after the name of the enumeration
%enum_select<Name{}0
// Missing { after the name of an enumerator
%enum_select<Name>{%OtherName|foo}0
// Enumerator name duplicates enumeration name? Maybe this is fine?
%enum_select<Name>{%Name{haha}}0
// No enumerators specified? Maybe this is fine?
%enum_select<Name>{foo|bar|baz}0
```

AaronBallman reviewed (commented):
Thank you for working on this! We should probably also update https://clang.llvm.org/docs/InternalsManual.html#formatting-a-diagnostic-argument so that folks know how to use this properly and we advertise it more broadly.

erichkeane committed:
Fix Review comments:

-Update internals-manual

-Update document generator comment, clarify we intend to document as if
it is a 'select'

-Add test coverage for other modifiers using '<'

-Add additional suggested test coverage

AaronBallman commented on clang/docs/InternalsManual.rst:
```diff
@@ -276,6 +276,21 @@ Description:
+  This format specifier is used exactly like a ``select`` specifier, except it
+  additionally generates a namespace, enumeration, and enumerator list based on
+  the format string given. In the above case, a namespace is generated named
+  ``FrobbleKind`` that has a anonymous enumeration with the enumerators
```
```suggestion
  ``FrobbleKind`` that has an unscoped enumeration with the enumerators
```

AaronBallman reviewed (approved):
LGTM with a minor nit on the documentation, but I think this is a great new feature for the diagnostics engine. Thank you for it!

btw, if you happen to have a list of diagnostics that could benefit from switching to use it, that sure seems like Good First Issue territory.

erichkeane committed:
Update clang/docs/InternalsManual.rst

Change anonymous to unscoped.  The name of it is irrelevant to what we're talking about anyway, since it is unscoped.

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

erichkeane commented:
> LGTM with a minor nit on the documentation, but I think this is a great new feature for the diagnostics engine. Thank you for it!
> 
> btw, if you happen to have a list of diagnostics that could benefit from switching to use it, that sure seems like Good First Issue territory.

Hmm... I don't.  I DID start walking down the list of diagnostics with 'large' selects at one point, but stopped at the first one I could find.  Perhaps I'll create a bug as a 'starter' bug with some guidance how to FIND these, and and hope someone comes along and does such an audit
